source/auth_credentials.c

```diff
@@ -120,10 +120,14 @@ PyObject *aws_py_credentials_session_token(PyObject *self, PyObject *args) {
  */
 struct credentials_provider_binding {
     struct aws_credentials_provider *native;
+    PyObject *py_provider;
 };
 
 /* Finally clean up binding (after capsule destructor runs and credentials provider shutdown completes) */
 static void s_credentials_provider_binding_clean_up(struct credentials_provider_binding *binding) {
+    if (binding->py_provider) {
+        Py_XDECREF(binding->py_provider);
+    }
```

Comment:
trivial: don't need the `if` here. Py_XDECREF does the if-check for you:

Py_DECREF: no if-check, always decrefs, will crash if variable is NULL
Py_XDECREF: if-check variable before decref
Py_CLEAR: if-check variable before decref, then sets variable to NULL